### PR TITLE
[Test] Make `test/test_cuda_trace.py` device-generic and enable testing XPU.

### PR DESCRIPTION
**Pull request by BBBela**
This PR refactors the `test/test_cuda_trace.py` to make it device-generic and enables testing XPU.

### Main Changes

- Rename `test_cuda_trace.py` -> `test_gpu_trace.py`.
- Change CODEOWNERS of this file from cuda-specific to generic as after refactor, this file is generic and uses common API calls.
- Rename test class `TestCudaTrace` -> `TestGpuTraceDevice`
- Instantiate `TestGpuTraceDevice` with `instantiate_device_type_tests`.
- Add `HardwareClassification`.
- Change importing hardcoded cuda module `import torch.cuda._gpu_trace as gpu_trace` to `importlib` approach based on the currently available accelerator. There is no device-generic `_gpu_trace` module, but several devices (cuda and xpu for sure) have the `_gpu_trace` module. By using `importlib` we can choose the right module based on the current accelerator.
- Change `torch.cuda.Event()` -> `torch.get_device_module(device).Event()`.
- Change `torch.cuda.default_stream().cuda_stream` -> `torch.accelerator.current_stream().native_handle`.
- Change hardcoded `"cuda"` to `device` parameter.
- Change `torch.cuda.Stream().cuda_stream` -> `torch.get_device_module(device).Stream().native_handle`
- Enable XPU testing.
- Add skips for XPU with linked issue for failing test cases.

### CODE REVIEW

**Comment by daisyden** [comment]
@liangan1 could you please help review this test port / refactor PR? Thanks!

**Comment by BBBela** [comment]
Hello @daisyden and @liangan1,

this PR isn't fully ready because I need some advice and decision before finalizing it. Here is the problem:

This test file contains API calls that can be generalized by using `torch.accelerator` or `torch.get_device_module(device)` or `Stream().native_handle`. Most of test cases (but not all) from this test file are already converted to xpu in `test_xpu.py` -> `TestXpuTrace` test class. This test file is listed under cuda codeowners and when I tried to change name from `test_cuda_[...]` to `test_gpu_[...]` the linter screams that I need also to change the codeowners entry.

So my questions is what should I do and what path to follow? Should we pursue making this file generic and changing codeowners from cuda to some generic owner (which one?) or create separate, new test file with those test cases but refactored to be generic and leave the cuda-hardcoded version as is? I am not sure and I need your decision regarding next steps. Maybe some even different solution would be best?

File: test/test_gpu_trace.py [hunk 1]
```diff
@@ -1,79 +1,84 @@
...
         gpu_trace.register_callback_for_memory_deallocation(self.mock)
 
-        tensor = torch.empty(3, 8, device="cuda")
+        tensor = torch.empty(3, 8, device=device)
         data_ptr = tensor.data_ptr()
         del tensor
         self.mock.assert_called_once_with(data_ptr)
 
-    def test_stream_creation_callback(self):
+    @onlyCUDA
```
Comment (liangan1) [inline test/test_gpu_trace.py:90]: Why only CUDA?

Comment (BBBela) [reply test/test_gpu_trace.py:90]: This test asserts cuda-specific behavior that is by design different on XPU.
The stream creation callback for CUDA is called when the `torch.cuda.Stream()` is called for the first time in the process. That is what this test asserts. For `ROCm`, the creation is even lazier, so it needs to run some operation on the stream to call the registered creation callback. That is why there is branching on `torch.version.hip` in the test.
For XPU, all the available streams are created eagerly, when `torch.xpu.current_device()` is called during test class setup, which is before the `torch.xpu.Stream()` call, and later on, the already created stream from the pool is used. So in a moment we call `torch.xpu.Stream()`, the streams are already created on XPU, so no creation callback is called.

That is why I marked this test as `@onlyCUDA`. I thought about moving it to a separate test class that is CUDA-specific, but for this one test, that fits well with the others, it seemed more logical.

And I used `@onlyCUDA` because this is not an XPU bug or problem that is going to be fixed, but rather design decision that simply is different for XPU than for CUDA. The other test is marked `@skipXPUIf` because a quick investigation showed that it is likely some missing feature or bug that can be fixed or at least needs further investigation.

File: test/test_gpu_trace.py [hunk 2]
```diff
@@ -1,79 +1,84 @@
...
-    print("CUDA not available, skipping tests", file=sys.stderr)
-    TestCase = NoTest
+
+if gpu := torch.accelerator.current_accelerator(check_available=True):
+    gpu_trace = importlib.import_module(f"torch.{gpu.type}._gpu_trace")
 
 
 @torch.testing._internal.common_utils.markDynamoStrictTest
-class TestCudaTrace(TestCase):
+class TestGpuTraceDevice(TestCase):
```
Comment (pbielak) [inline test/test_gpu_trace.py:33]: Please add `hw_classification` to the class

Comment (BBBela) [reply test/test_gpu_trace.py:33]: Right, thanks for this catch! Added!

**Review by pbielak — APPROVED** [review]

**Comment by BBBela** [comment]
[EDITED]
~~I see new problems and failures in the CI. Please do not review, until I fix those problems. I will comment once PR is ready for review.~~
It is ready for review now.

**Comment by liangan1** [comment]
LGTM

**Review by liangan1 — APPROVED** [review]

File: test/test_gpu_trace.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    def test_memory_deallocation_callback(self, device):
+        gpu_trace.register_callback_for_memory_deallocation(self.mock)
+
+        tensor = torch.empty(3, 8, device=device)
+        data_ptr = tensor.data_ptr()
+        del tensor
+        self.mock.assert_called_once_with(data_ptr)
+
+    @onlyCUDA
+    @skipCUDANonDefaultStreamIf(True)
```
Comment (BBBela) [inline test/test_gpu_trace.py:91]: It was necessary to add the `@skipCUDANonDefaultStreamIf(True)` decorator here.
The original approach, where test class inherited directly from `TestCase`, was using default `_do_cuda_non_default_stream = False`. Because of that setting, the stream creation was being created when the below test executed `torch.cuda.Stream()` as it was first stream creation. Then the callback for stream creation was called. When I switched to `instantiate_device_type_tests` approach, the test class for CUDA inherits from `CUDATestBase` that sets the `_do_cuda_non_default_stream = True`. That makes the streams initialization before entering the test function. At the moment the `torch.cuda.Stream()` is called, the pool of streams is already created and creation callback has already been called.

Saying that, I used the `@skipCUDANonDefaultStreamIf(True)` decorator that effectively switches `_do_cuda_non_default_stream` back to False and allows correct test behavior and first streams creation when the `torch.cuda.Stream()` is called.

**Comment by BBBela** [comment]
Hi @pbielak, there were some new changes introduced since your last review - please re-review and let me know if everything is okay or if you have any additional comments or suggestions?

**Review by pbielak — APPROVED** [review]

**Comment by BBBela** [comment]
Added small refactor in latest commit **"Move importlib into test class and parametrize lib based on self.device_type"**

File: test/test_gpu_trace.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
+# Owner(s): ["module: tests"]
```
Comment (etaf) [inline test/test_gpu_trace.py]: Changing the owner from `module: cuda` to `module: tests` sends disabled-test and flaky-test issues for these tests to test infra instead of the people who own GPU trace. The other files ported with `only_for=("cuda", "xpu")` keep their feature owner (dynamo, inductor, profiler, ...). I'd keep `module: cuda` here, and add an XPU label next to it if you want XPU failures routed to you too.

File: test/test_gpu_trace.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    def test_memory_deallocation_callback(self, device):
+        self.gpu_trace.register_callback_for_memory_deallocation(self.mock)
+
+        tensor = torch.empty(3, 8, device=device)
+        data_ptr = tensor.data_ptr()
+        del tensor
+        self.mock.assert_called_once_with(data_ptr)
+
+    @onlyCUDA
+    @skipCUDANonDefaultStreamIf(True)
```
Comment (etaf) [inline test/test_gpu_trace.py]: Your explanation in the thread of why `@skipCUDANonDefaultStreamIf(True)` is needed (CUDATestBase builds the stream pool in its non-default-stream wrapper before the test body runs, so the creation callback never fires) isn't obvious from the code. Someone cleaning this up later will likely delete the decorator and break the test. Could you add a one-line comment here?

Also, since the test is `@onlyCUDA` and the HIP branch already calls `torch.cuda.Stream()`, line 95 can just be `torch.cuda.Stream()` too. Using `torch.get_device_module(device)` only in the else branch makes it look like the test is meant to be device-generic when it isn't.

**Review by etaf — APPROVED** [review]
